nexus/src/external_api/console_api.rs

```diff
@@ -546,40 +490,73 @@ pub struct LoginUrlQuery {
-    // we can look at the cookie and default to sending them to the IdP
-    // indicated (though if they don't want that one we need to make sure they
-    // can get to a different one). If there is no cookie, we send them to the
-    // selector page. In any case, none of this is done here yet. We go to
-    // /spoof_login no matter what.
-    let login_uri = "/spoof_login";
+/// Generate URI to the appropriate login form for this Silo. Optional
+/// `redirect_url` represents the URL to send the user back to after successful
+/// login, and is included in `state` query param if present
+async fn get_login_url(
+    rqctx: &RequestContext<Arc<ServerContext>>,
+    redirect_url: Option<String>,
+) -> Result<String, Error> {
+    let nexus = &rqctx.context().nexus;
+    let endpoint = nexus.endpoint_for_request(rqctx)?;
+    let silo = endpoint.silo();
+
+    let login_uri = if silo.authentication_mode == AuthenticationMode::Local {
+        format!("/login/{}/local", silo.name())
+    } else {
+        // List the available identity providers and pick an arbitrary one.
+        // It might be nice to have some policy for choosing which one is used
+        // here.
+        let opctx = nexus.opctx_external_authn();
+        let silo_lookup = nexus.silo_lookup(opctx, NameOrId::Id(silo.id()))?;
+        let idps = nexus
+            .identity_provider_list(
+                opctx,
+                &silo_lookup,
+                &PaginatedBy::Name(DataPageParams {
+                    marker: None,
+                    direction: dropshot::PaginationOrder::Ascending,
+                    limit: NonZeroU32::try_from(2).unwrap(),
+                }),
+            )
+            .await?;
+
+        if idps.is_empty() {
+            let error = Error::invalid_request(
+                "no identity providers are configured for Silo",
+            );
+            warn!(&rqctx.log, "{}", error; "silo_id" => ?silo.id());
+            return Err(error);
+        }
+
+        if idps.len() > 1 {
+            warn!(
+                &rqctx.log,
+                "found more than one IdP for Silo";
+                "silo_id" => ?silo.id()
+            );
+        }
+
+        format!(
+            "/login/{}/saml/{}",
+            silo.name(),
+            idps.into_iter().next().unwrap().name()
+        )
+    };
 
     // Stick redirect_url into the state param and URL encode it so it can be
     // used as a query string. We assume it's not already encoded.
     let query_data = LoginUrlQuery { state: redirect_url };
 
-    match serde_urlencoded::to_string(query_data) {
+    Ok(match serde_urlencoded::to_string(query_data) {
         // only put the ? in front if there's something there
         Ok(encoded) if !encoded.is_empty() => format!("{login_uri}?{encoded}"),
         // redirect_url is either None or not url-encodable for some reason
-        _ => login_uri.to_string(),
-    }
+        _ => login_uri,
+    })
```

Comment:
Thinking out loud here, so bear with me. The `state` param will not work for SAML login, at least not directly. It was designed for local login: when the login POST comes back successful, the [console pulls the param from the URL](https://github.com/oxidecomputer/console/blob/1eb450f0a382cf5308b627c3597d7459cd3eb5cb/app/pages/LoginPage.tsx#L34) and navigates to it if present. For SAML login we have to send the state to the IdP so they can send it back to us on success. We do that with `RelayState`, which is a free-form thing — you can put anything in there. We have a referer param (analogous to `state`) and we only base64 it right now, but we will probably eventually want to sign it too, which means the console cannot construct it.

https://github.com/oxidecomputer/omicron/blob/8d32437da0fe9a56c63929c5a70083ebfeff7982/nexus/src/external_api/console_api.rs#L210-L236

For the SAML login page, I figured there would be a link button that goes to the IdP. Because we need the server to encode the redirect URL into `RelayState` (assuming we want to plan for a future where the thing is signed), this button would have to go to Nexus, including the `state` param in the URL again, and Nexus would return a redirect to the SAML IdP. Alternatively (and more or less equivalent) instead of hitting Nexus to redirect to the IdP, when rendering the page client-side we could make an API call that gives us the direct IdP URL and we stick that in the link `href`.

Here is a way it _could_ work, which is a bit silly and roundabout, but it's a straightforward consequence of the fact that we are not doing any console templating server-side.

1. User hits a 401 and gets send to `/login?state=/path/I/wanted`
2. `/login` determines we should use SAML and returns a redirect to `/login/my-silo/saml/my-idp?state=/path/I/wanted`
3. This is a "console page" so we render the console and show a page with a button that says "Go to IdP" or whatever
4. That button is a link to `/login/my-silo/saml/my-idp/direct?state=/path/I/wanted`, and when clicked, nexus responds with a 302 redirect to `my-idp.com/login/whatever?RelayState=kdfjhsdjdks54d565`

Everything after this point is already working, I think — after successful login, read the `referer` out of `RelayState` and redirect there.

Like I said, this feels silly because it requires a whole other endpoint in Nexus to turn `state` into `RelayState`. If we were trying to be Normal and render the login page server-side, we would be able to do that in the handler for `/login/my-silo/saml/my-idp`. I do often wonder when we're going to hit the point where it becomes worth it to do some server-side HTML templating instead of doing things like this, but having the console frontend continue to be a hermetically sealed box still seems safest to me. We could also punt on that and let console construct `RelayState` itself, but again that will stop working once we want to sign it.

Reply:
See draft console PR for SAML login page here:
https://github.com/oxidecomputer/console/pull/1549

Reply:
While writing that PR, I thought about what words to put on the page to tell the user what is going on, and it occurred to me that that's an advantage of the approach where console makes an API call to get the complete IdP URL — it would let us include some other info like a name or description of the IdP.

<img width="1023" alt="image" src="https://github.com/oxidecomputer/omicron/assets/3612203/8cdd77ae-ddf3-4c41-ac5b-3586ed5e1796">

Reply:
I think that's all reasonable, and also stuff I'd like to defer for now.  I think most of this PR is a prereq for that change anyway.  That okay?

Reply:
Yeah, I think it’s good to merge.